### PR TITLE
Add WeakRecipient (v2)

### PR DESCRIPTION
**Pull request by hoodie**
With this change you can now call `.recipient()` on `WeakAddr<A>` just
like on `Addr<A>`. In this case you get a `WeakRecipient<M>` that can
be upgraded to a `Recipient<M>`.

## PR Type

INSERT_PR_TYPE

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview
Hi there,
I'm trying to convert my project to make greater use of `Recipient<M>` instead of `Addr<A>` since I believe it allows to loosen some tight coupling here and there. Unfortunately there is no equivalent to `WeakAddr<A>` for Recipients yet, so I cannot really use Recipients at the moment without running the risk of leaking objects.
Therefore I would like to suggest adding `WeakRecipients` to actix.

This PR supersedes #418, which was flawed. In that attempt I made the `Box` inside the `Recipient` an `Arc` and downgraded it.
you would call
```rust
let weak_rcp = my_actor.addr().recipient().downgrade();
```

That however meant  that you the `WeakRecipient` was invalidated immediately.
In this second attempt I'm creating the `WeakRecipient` from an existing `WeakAddr`. 
```rust
let weak_rcp = my_actor.addr().downgrade().recipient();
```

Thank you very much for your review.

### CODE REVIEW

**Comment by hoodie** [comment]
@JohnTitor @robjtede ?

File: src/address/mod.rs [hunk 1]
```diff
@@ -329,6 +339,58 @@ where
...
+    A: Handler<M>,
+    M::Result: Send,
+    A::Context: ToEnvelope<A, M>,
+{
+    fn into(self) -> WeakRecipient<M> {
+        WeakRecipient::new(Box::new(self.wtx))
+    }
+}
+
+/// The `Recipient` type allows to send one specific message to an
```
Comment (lucab) [inline src/address/mod.rs:388]: Spurious docstring (possibly a copy-paste leftover)?

File: src/address/channel.rs [hunk 2]
```diff
@@ -36,6 +36,13 @@ where
 
     fn connected(&self) -> bool;
 }
+pub trait WeakSender<M>: Send
```
Comment (lucab) [inline src/address/channel.rs:39]: Would `pub(crate)` be enough here?

File: src/address/mod.rs [hunk 3]
```diff
@@ -329,6 +339,58 @@ where
     }
 }
 
+/// A weakly referenced counterpart to `Recipient<M>`
```
Comment (lucab) [inline src/address/mod.rs:342]: As a user, I'd find some more docstring helpful here to explain what's the typical usecase of this weak recipient.
In particular, I found your mention of "risk of leaking objects" enlightening; it could mirrored and expanded in the docstring here.

**Review by lucab — COMMENTED** [review]
(just some casual PR reading, I'm not a maintainer here)

File: src/address/mod.rs (hunk 1 above)
Comment (hoodie) [reply src/address/mod.rs:388]: whoops, yes indeed, I was shuffling files around a bit there, after I had moved recipient and weak_recipient into their own submodules and then reintegrated them. these files contain a bunch of structs at once and tons of trait implementations, I would at some point ask if we perhaps want to split them up a little.

tanks for the find!

File: src/address/mod.rs (hunk 3 above)
Comment (hoodie) [reply src/address/mod.rs:342]: this doc-string is a reflection of the one for `WeakAddr`. You are right about the usecase example, IDK if it fits in here, but maybe an example or integration test would be good.

**Comment by hoodie** [comment]
Ping @JohnTitor 😀

**Review by JohnTitor — COMMENTED** [review]
Hey, sorry for the delay, and thanks for pinging!
The approach should be reasonable and the change looks good to me. But I'm going to cc @Jonathas-Conceicao for further reviewing, if you're available now :)
Also, could you add an entry to the changelog?

File: actix-derive/src/message.rs [hunk 4]
```diff
@@ -69,10 +69,10 @@ fn get_attribute_type_multiple(
             .map(|m| meta_item_to_ty(m).ok())
             .collect())
     } else {
-        return Err(syn::Error::new_spanned(
+        Err(syn::Error::new_spanned(
             attr,
             format!("The correct syntax is #[{}(type, type, ...)]", name),
-        ));
+        ))
```
[the remark this answers is not in the document]
Comment (hoodie) [reply actix-derive/src/message.rs:75]: sure, will do 👍

File: src/address/mod.rs (hunk 3 above)
Comment (JohnTitor) [reply src/address/mod.rs:342]: The docs enhancement isn't absolutely necessary, but it'd be great if we could add a minimal example for it :)

**Comment by hoodie** [comment]
@robjtede why did you merge `master` into this branch?

**Comment by robjtede** [comment]
@hoodie repo has a linear history policy, so if this is accepted it needs to be based on HEAD for CI to check it properly before squashing

**Comment by hoodie** [comment]
ah ok, so merging master in here is a actually better than rebasing onto master for this use case, get it

**Comment by robjtede** [comment]
@hoodie either is appropriate and just as good, many people don't like rebasing and it's just a button on github to merge master in

File: src/address/channel.rs [hunk 5]
```diff
@@ -555,6 +562,22 @@ impl<A: Actor> WeakAddressSender<A> {
...
 }
 
+impl<A, M> WeakSender<M> for WeakAddressSender<A>
+where
+    A: Handler<M>,
+    A::Context: ToEnvelope<A, M>,
+    M::Result: Send,
+    M: Message + Send + 'static,
+{
+    fn upgrade(&self) -> Option<Box<dyn Sender<M> + Sync>> {
```
Comment (Jonathas-Conceicao) [inline src/address/channel.rs:572]: I believe we should have a doc note about when this `Option` would be `None`, either here or at the trait's definition.

File: tests/test_address.rs [hunk 6]
```diff
@@ -100,7 +101,51 @@ impl Actor for WeakRunner {
...
+
+        let weak1 = addr1.downgrade().recipient();
+        let weak2 = addr2.downgrade().recipient();
+        // not strictly necessary, weak1 is dropped by RIAA since nobody in `run_later` uses it.
+        // to make this test red, move drop(addr1) to the line before `System::current().stop()`
+        drop(addr1);
+
+        ctx.run_later(Duration::new(0, 1000), move |_, _| {
+            if weak1.upgrade().is_none() {
+            } else {
```
Comment (Jonathas-Conceicao) [inline tests/test_address.rs:130]:
```suggestion
            if weak1.upgrade().is_some() {
```

**Review by Jonathas-Conceicao — CHANGES_REQUESTED** [review]
The feature proposal seams good overall, I left a couple of comments for some minor improvements.

File: src/address/channel.rs (hunk 5 above)
Comment (hoodie) [reply src/address/channel.rs:572]: sure, but probably at the definition of the trait and not here right?

Comment (Jonathas-Conceicao) [reply src/address/channel.rs:575]: Both will be generated to the doc, but I think the trait it might be more noticeable. So if the `None` has a generic cause it would be better to have it there, if it's specific to the type's implementation we can have it here.

Comment (hoodie) [reply src/address/channel.rs:575]: I added it on the trait, but it's `pub(crate)`, so it won't show up in any docs anyway.
Plus: the semantics are basically the same as for `Weak<T>::upgrade()`.
Alright by you @Jonathas-Conceicao ?

Comment (hoodie) [reply src/address/channel.rs:575]: I added it on the trait, but it's pub(crate), so it won't show up in any docs anyway.
Plus: the semantics are basically the same as for Weak<T>::upgrade().
Alright by you @Jonathas-Conceicao ?

**Comment by hoodie** [comment]
@JohnTitor I have added an example for `WeakAddr` and `WeakRecipient`.
In the process I noticed that the signature of `Recipient::do_send()` is not consistent with the signature of `Addr::do_send()`, I believe the former should not return a `Result<...>`. I just opened #440 to track that.

**Comment by hoodie** [comment]
Ok, since this has gotten a bunch of changes since I opened it originally I thought I'd give a quick recap since I know it's a bit hard to follow these things:

* [x] added `WeakRecipient`
* [x] added trait doc for `WeakAddressSender`
* [x] added two new examples
* [ ] make clippy happy

any other requests?

File: src/address/mod.rs (hunk 3 above)
Comment (lucab) [reply src/address/mod.rs:342]: @hoodie thanks for the pair of examples, much appreciated!

**Comment by hoodie** [comment]
ping @Jonathas-Conceicao

**Review by Jonathas-Conceicao — APPROVED** [review]
lgtm, thanks for working on this!

**Comment by JohnTitor** [comment]
Thanks all!
